feat: rework architecture page (#116)
ZephyrCloudIO/zephyr-documentation: docs/learning/concepts/architecture.mdx

---
- title: Architecture
- description: Technical architecture and implementation details of Zephyr
---

# Architecture

> Zephyr is a distributed edge deployment platform that integrates directly with JavaScript bundlers to provide instant deployment and global distribution of web applications.

Reviewer: ```suggestion
Zephyr is a distributed edge deployment platform that integrates directly with JavaScript build tools including but not limited to `rspack`, `webpack`, `rollup`, `rolldown`, `vite`, `repack` and soon `metro` - to provide instant deployment and global distribution of web applications.
```

because bundlers are included in build tools - higher scope

This document details the technical architecture, design decisions, and implementation strategies that enable sub-second deployments and millisecond-latency serving.

The platform achieves its performance characteristics through a combination of compile-time integration, intelligent asset processing, and a globally distributed edge network.

By intercepting the build process at the bundler level, Zephyr can analyze, optimize, and deploy applications without modifying the build output or requiring changes to existing workflows.

## Core Architecture Principles

Zephyr's architecture is built on several key principles that drive its design:

**Non-invasive Integration**: The platform integrates through plugins that preserves existing bundler configurations while adding deployment capabilities.

```ts title=rspack.config.ts
export default withZephyr()({
  context: __dirname,
  entry: {
    main: './src/main.jsx',
  },
  // existing configuration remains unchanged
});
```

**Edge-first Design**: All serving infrastructure operates at the edge, utilizing a global network of workers to minimize latency and maximize availability.

> **Content-addressed Storage**: Assets are stored and retrieved using content-based addressing, enabling perfect caching, deduplication, and atomic deployments.

Reviewer: ```suggestion
**Content-addressed Storage**: Assets are stored and retrieved using content-based addressing, enabling perfect caching, deduplication, and atomic deployments.

While most of these features, capabilities can be available under customer's own infrastructure if configured (configuration is simple btw). 
```

## System Architecture Overview

![Zephyr Infrastructure Architecture](/infra-overview.png)

The Zephyr platform consists of four primary subsystems that work together to provide seamless deployment and serving capabilities:

- **Build Integration Layer**: The entry point for developers, implemented as plugins for popular bundlers (Webpack, Rspack, Vite, etc...).

>   These plugins intercept build lifecycle events to analyze outputs, extract metadata, and initiate the deployment process. The integration is designed to be transparent, adding no overhead to the build process and requiring no changes to existing configurations.

Reviewer: ```suggestion
  These plugins intercept build lifecycle events to analyze outputs, extract metadata, and initiate the deployment process. The integration is designed to be transparent, adding no overhead to the build process and requiring no changes to existing configurations. You can see a list of existing plugins and their usages [here](https://docs.zephyr-cloud.io/recipes). 
```

so we have backlinks

> - **Control Plane**: The centralized management system comprising the API platform and dashboard UI.

Reviewer: ```suggestion
- **Control Plane**: The centralized management system comprising the Zephyr Cloud's API platform and dashboard UI.
```

Author: Added

  It handles authentication, authorization, project management, and deployment orchestration. The control plane maintains the source of truth for all deployments and provides both programmatic and visual interfaces for management.

> - **Data Plane**: The distributed edge infrastructure responsible for serving applications.

Reviewer: ```suggestion
- **Data Plane**: The distributed edge infrastructure responsible for serving applications either on our own cloud or on customer's cloud. 
```
Still need better framing for this
